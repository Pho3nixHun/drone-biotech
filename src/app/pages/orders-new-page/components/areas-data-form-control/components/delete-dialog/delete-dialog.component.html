@if (vm(); as vm) {
    <app-dialog-layout
        class="w-80 md:w-96 block"
        (close)="cancelClick()"
        [vm]="vm"
        *transloco="let t"
    >
        <h3 class="text-lg">
            {{ t(vm.warningTextKey) }}
        </h3>
        <app-button [vm]="vm.deleteButtonVM" (clicked)="deleteClick()" />
        <app-button [vm]="vm.cancelButtonVM" (clicked)="cancelClick()" />
    </app-dialog-layout>
}
