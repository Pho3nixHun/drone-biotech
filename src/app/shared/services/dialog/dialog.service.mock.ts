/* eslint-disable @typescript-eslint/no-unused-vars */
import { Injectable } from '@angular/core';
import { noop, Subject } from 'rxjs';
import { DialogRef } from './dialog.service';

@Injectable({
    providedIn: 'any',
})
export class DialogServiceMock {
    public create(_vm: unknown, _component: unknown): DialogRef {
        return {
            result$: dialogResponse.asObservable(),
            close: () => noop,
            open: () => noop,
            onClose: () => noop,
            result: () => noop,
            portal: undefined,
            resultSubject$: new Subject(),
        } as unknown as DialogRef;
    }
}

// Using Subject instead of signal to avoid requiring an initial value
const dialogResponse = new Subject<unknown>();

export const updateDialogResponse = (response: unknown) =>
    dialogResponse.next(response);
