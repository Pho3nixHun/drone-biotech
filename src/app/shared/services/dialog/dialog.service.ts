/* eslint-disable @typescript-eslint/no-explicit-any */
import {
    GlobalPositionStrategy,
    Overlay,
    OverlayRef,
    ScrollStrategy,
} from '@angular/cdk/overlay';
import { ComponentPortal, Portal } from '@angular/cdk/portal';
import { Injectable, InjectionToken, Injector, inject } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable, Subject, map, merge, skip } from 'rxjs';
import { selectURL } from 'src/app/stores/router/router.selectors';

export interface DialogReason {
    reasonType: 'submit' | 'cancel';
}
export class DialogRef<T = unknown> {
    private portal: Portal<unknown> | undefined;
    private readonly resultSubject$ = new Subject<T | void>();
    constructor(public onClose: (result?: T) => void) {}

    public open(portal: Portal<unknown>, overlayRef: OverlayRef) {
        this.portal = portal;
        if (!overlayRef?.hasAttached()) {
            overlayRef?.attach(this.portal);
        }
    }

    public close(result?: T): void {
        this.resultSubject$.next(result);
        this.resultSubject$.complete();
        this.onClose(result);
    }

    public set result(value: T) {
        this.resultSubject$.next(value);
    }

    public get result$(): Observable<T | void> {
        return this.resultSubject$.asObservable();
    }
}

export const DIALOG_DATA = new InjectionToken<any>('DIALOG_DATA');
export const DIALOG_REF = new InjectionToken<DialogRef>('DIALOG_REF');

@Injectable({
    providedIn: 'root',
})
export class DialogService {
    private readonly overlay = inject(Overlay);
    private readonly store = inject(Store);

    private readonly routerEvent$ = this.store.select(selectURL);
    private readonly overlayRef = this.overlay.create({
        hasBackdrop: true,
        positionStrategy: this.overlay
            .position()
            .global()
            .centerHorizontally()
            .centerVertically(),
        scrollStrategy: this.overlay.scrollStrategies.block(),
    });

    create<VM, R = any>(
        viewModel: VM,
        component: any,
        scrollStrategy?: ScrollStrategy,
        positionStrategy?: GlobalPositionStrategy,
        close$: Observable<void> = this.routerEvent$.pipe(
            skip(1),
            map(() => void 0)
        )
    ): DialogRef<R> {
        if (positionStrategy) {
            this.overlayRef.updatePositionStrategy(positionStrategy);
        }
        if (scrollStrategy) {
            this.overlayRef.updateScrollStrategy(scrollStrategy);
        }
        const dialogRef = new DialogRef<R>(() => {
            this.overlayRef?.detach();
        });

        const injector = Injector.create({
            providers: [
                {
                    provide: DIALOG_DATA,
                    useValue: viewModel,
                },
                {
                    provide: DIALOG_REF,
                    useValue: dialogRef,
                },
            ],
        });
        const dialogPortal = new ComponentPortal(component, null, injector);

        dialogRef.open(dialogPortal, this.overlayRef);

        merge(this.overlayRef?.backdropClick(), close$).subscribe(() => {
            dialogRef.close();
        });

        return dialogRef;
    }
}
