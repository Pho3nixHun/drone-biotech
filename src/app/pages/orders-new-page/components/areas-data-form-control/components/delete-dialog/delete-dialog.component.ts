import { TranslocoModule } from '@jsverse/transloco';
import { MatIconModule } from '@angular/material/icon';
import { Component, computed, inject, signal } from '@angular/core';
import { DIALOG_DATA, DIALOG_REF } from '@services/dialog/dialog.service';
import { DeleteDialogReason, isDeleteDialogVM } from './delete-dialog.model';
import { DialogLayoutComponent } from '@components/dialog-layout/dialog-layout.component';
import { ButtonComponent } from '@components/button/button.component';

@Component({
    selector: 'app-delete-dialog',
    imports: [
        MatIconModule,
        TranslocoModule,
        DialogLayoutComponent,
        ButtonComponent,
    ],
    templateUrl: './delete-dialog.component.html',
})
export class DeleteDialogComponent {
    protected readonly unknownVM = signal<unknown>(inject(DIALOG_DATA));
    protected readonly dialogRef = inject(DIALOG_REF);

    protected vm = computed(() => {
        const vm = this.unknownVM();
        return isDeleteDialogVM(vm) ? vm : null;
    });

    protected deleteClick() {
        const dialogCloseReason: DeleteDialogReason = {
            reasonType: 'submit',
            type: 'deleteDialogReason',
        };
        this.dialogRef.close(dialogCloseReason);
    }

    protected cancelClick() {
        const dialogCloseReason: DeleteDialogReason = {
            reasonType: 'cancel',
            type: 'deleteDialogReason',
        };
        this.dialogRef.close(dialogCloseReason);
    }
}
