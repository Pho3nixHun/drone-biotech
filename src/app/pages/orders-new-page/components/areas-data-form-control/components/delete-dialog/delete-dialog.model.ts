import { ButtonVM } from '@components/button/button.model';
import { DialogLayoutVM } from '@components/dialog-layout/dialog-layout.model';
import { DialogReason } from '@services/dialog/dialog.service';
import { isObject } from '@utils/is-object.typeguard';

export interface DeleteDialogVM extends DialogLayoutVM {
    type: 'deleteDialogVM';
    titleKey: string;
    warningTextKey: string;
    deleteButtonVM: ButtonVM;
    cancelButtonVM: ButtonVM;
}

export interface DeleteDialogReason extends DialogReason {
    type: 'deleteDialogReason';
}

export const isDeleteDialogResult = (
    data: unknown
): data is DeleteDialogReason => {
    return (
        isObject(data) && 'type' in data && data.type === 'deleteDialogReason'
    );
};

export function isDeleteDialogVM(data: unknown): data is DeleteDialogVM {
    return isObject(data) && 'type' in data && data.type === 'deleteDialogVM';
}
