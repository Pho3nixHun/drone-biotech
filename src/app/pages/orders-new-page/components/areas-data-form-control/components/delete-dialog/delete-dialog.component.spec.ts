import { ComponentFixture, TestBed } from '@angular/core/testing';
import { DeleteDialogComponent } from './delete-dialog.component';
import { DeleteDialogVM } from './delete-dialog.model';
import { getTranslocoModule } from 'transloco-testing.module';
import { DIALOG_REF, DialogRef } from '@services/dialog/dialog.service';
import {
    provideMockDialogDataFactory,
    provideMockDialogRef,
    updateDialogDataSignal,
} from 'src/app/shared/providers/dialog-provider';
import { MatIcon } from '@interfaces/mat-icon.enum';

const enMock = {
    title: 'tit',
    warningText: 'warning',
};

describe('DeleteDialogComponent', () => {
    let fixture: ComponentFixture<DeleteDialogComponent>;
    let compiled: HTMLElement;
    let component: DeleteDialogComponent;
    const vm: DeleteDialogVM = {
        cancelButtonVM: {
            type: 'withIcon',
            icon: MatIcon.ADD,
            variant: 'fill',
        },
        closeButtonVM: {
            type: 'withIcon',
            icon: MatIcon.ADD,
            variant: 'fill',
        },
        deleteButtonVM: {
            type: 'withIcon',
            icon: MatIcon.ADD,
            variant: 'fill',
        },
        titleKey: enMock.title,
        type: 'deleteDialogVM',
        warningTextKey: enMock.warningText,
    };

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                DeleteDialogComponent,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [provideMockDialogDataFactory(), provideMockDialogRef()],
        }).compileComponents();
    });

    describe('when the vm is not correct', () => {
        beforeEach(() => {
            updateDialogDataSignal({});
            fixture = TestBed.createComponent(DeleteDialogComponent);
            compiled = fixture.debugElement.nativeElement;
        });

        // Snapshot testing
        it('should not render the content if not DeleteDialogVM is provided', () => {
            // Arrange

            // Act
            fixture.detectChanges();

            //Assert
            expect(compiled).toMatchSnapshot();
        });
    });
    describe('when the vm is correct', () => {
        let dialogRef: DialogRef<unknown>;
        beforeEach(() => {
            updateDialogDataSignal(vm);
            fixture = TestBed.createComponent(DeleteDialogComponent);
            dialogRef = TestBed.inject(DIALOG_REF);
            compiled = fixture.debugElement.nativeElement;
        });

        // Snapshot testing
        it('should render the content', () => {
            // Arrange

            // Act
            fixture.detectChanges();

            //Assert
            expect(compiled).toMatchSnapshot();
        });

        // Unit testing
        it('should close the dialog when the deleteClick() method is called', () => {
            // Arrange
            const deleteSpy = jest.spyOn(dialogRef, 'close');
            component = fixture.componentInstance;

            // Act
            component['deleteClick']();

            //Assert
            expect(deleteSpy).toHaveBeenCalled();
        });

        // Unit testing
        it('should close the dialog when the cancelClick() method is called', () => {
            // Arrange
            const deleteSpy = jest.spyOn(dialogRef, 'close');
            component = fixture.componentInstance;

            // Act
            component['cancelClick']();

            //Assert
            expect(deleteSpy).toHaveBeenCalled();
        });
    });
});
