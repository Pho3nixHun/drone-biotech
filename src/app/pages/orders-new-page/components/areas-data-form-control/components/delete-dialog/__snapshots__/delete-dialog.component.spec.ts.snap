// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DeleteDialogComponent when the vm is correct should render the content 1`] = `
<div
  id="root1"
  ng-version="20.1.7"
>
  <app-dialog-layout
    class="w-80 md:w-96 block"
  >
    <section
      class="p-4 md:p-10 rounded-lg shadow-md transition-all duration-100 bg-white space-y-5 flex flex-col"
    >
      <div
        class="flex justify-between"
      >
        <h3
          class="text-3xl"
        >
           tit 
        </h3>
        <app-button>
          <button
            class="btn btn-primary btn-square"
          >
            <!--container-->
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
               add 
            </mat-icon>
            <!--container-->
          </button>
        </app-button>
      </div>
      <div
        class="space-y-4"
      >
        <h3
          class="text-lg"
        >
           warning 
        </h3>
      </div>
      <div
        class="ml-auto join"
      >
        <app-button>
          <button
            class="btn btn-primary btn-square"
          >
            <!--container-->
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
               add 
            </mat-icon>
            <!--container-->
          </button>
        </app-button>
        <app-button>
          <button
            class="btn btn-primary btn-square"
          >
            <!--container-->
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
               add 
            </mat-icon>
            <!--container-->
          </button>
        </app-button>
      </div>
    </section>
    <!--container-->
  </app-dialog-layout>
  <!--container-->
  <!--container-->
</div>
`;

exports[`DeleteDialogComponent when the vm is not correct should not render the content if not DeleteDialogVM is provided 1`] = `
<div
  id="root0"
  ng-version="20.1.7"
>
  <!--container-->
</div>
`;
